### PR TITLE
Simpler and safe implementation with improved UX

### PR DESCRIPTION
**Pull request by Gabriella439**
This is essentially two changes bundled into one (and I'll explain in just a second why I bundled them)

The first change was to remove the dependence on `unsafeCoerce` by changing the internal implementation to use an implicit parameter. In fact, this is literally the exact same way `HasCallStack` is implemented except without the GHC magic to automatically satisfy the constraint if it is omitted.  In other words, all that the `require-callstack` package does is implement `HasCallStack` without the magic so that the constraint has to be explicitly propagated.

One of the benefits of this approach is that now users can also satisfy the constraint by just defining the missing implicit parameter:

```haskell
let ?provideCallStack = ()
```

… but I kept the existing `provideCallStack` function both for backwards compatibility and also for the benefit of users who don't want to enable `ImplicitParams` (which is probably a lot of them, to be honest).

The second change was to switch to using `TypeError` to improve the UX instead of a cleverly named implicit parameter like `?add_RequireCallStack_to_function_context_or_use_provideCallStack`. The reason I bundled this change was because the choice of whether or not to use `TypeError` affects what to name the implicit parameter. I wanted to avoid having to rename the implicit parameter if we were to switch to `TypeError` later since the implicit parameter's name would technically become part of our public API if we were to switch to this new approach (since users could begin defining the implicit parameter instead of using `provideCallStack`).

So I did some research to figure out if it was feasible to switch to `TypeError` right now and the answer is: yes!  Specifically, I tested both `TypeError` and `Unsatisfiable` and learned the following things:

- `TypeError` actually works fine if you add type annotations to all intermediate `let` helper bindings that wrap `errorRequiredCallStack`

- GHC 9.8's support for `Unsatisfiable` doesn't actually change the design space at all

  Both `TypeError` and `Unsatisfiable` have the exact same problem where they trigger premature type errors if you define intermediate `let` helper bindings without a type annotation.

So since 9.8 isn't going to change the design space, I decided to see if it was realistic to just ask people to annotate intermediate `let` bindings that use `errorCallStack`.  Specifically, I tested this PR against our internal codebase and found that it required minimal changes to adopt this new `TypeError` approach.

This change only required adding 11 type annotations across 3 modules (out of 6000+ modules), so I inferred from this that it wouldn't be significantly disruptive to end users to switch to this new `TypeError`-based UX.

Moreover, this change actually slightly improved GHC's constraint resolution.  After making these two changes GHC now found two redundant `HasCallStack` constraints within our codebase, whereas the previous approach was somehow confounding GHC's ability to detect those gratuitous constraints.

### CODE REVIEW

File: src/RequireCallStack.hs [hunk 1]
```diff
@@ -48,9 +65,50 @@ import RequireCallStack.Internal
...
+type RequireCallStackImpl = ?provideCallStack :: ()
+
+-- | Raise an 'Control.Exception.ErrorCall' and incur a 'RequireCallStack'
+-- constraint while you do so. This
 --
 -- @since 0.1.0.0
 errorRequireCallStack :: RequireCallStack => String -> x
 errorRequireCallStack = error
+
+instance TypeError ('Text "Add RequireCallStack to your function context or use provideCallStack") => IP "provideCallStack" ()
```
Comment (parsonsmatt) [inline src/RequireCallStack.hs]: Oh that is very nice!

**Review by parsonsmatt — COMMENTED** [review]
This is great, thanks! Removing the need for `unsafeCoerce` is always a blessing.

I'm curious how the `ImplicitParams` nature of this plays out. Do end users "see through' the type alias and end up with an error message mentioning `ImplicitParams`? 

I think I'd rather keep the `ImplicitParams` implementation detail hidden from end users - I think we could do that by defining a datatype that isn't exported as the value of the implicit param, and only `provideCallStack` actually uses it. Sort of like,

```haskell
data HiddenToken = HiddenToken

type RequireCallStackImpl = (?provideCallStack :: HiddenToken)
```

What do you think of that?

File: src/RequireCallStack.hs (hunk 1 above)
Comment (parsonsmatt) [reply src/RequireCallStack.hs]: Wait, at what point is this actually invoked?

**Comment by Gabriella439** [comment]
End users won't ever see an error message mentioning the implicit because of the `TypeError` stuff so the implicit parameter won't show up in inferred types or error messages.  Importantly, if the user tries to infer a type involving the implicit param they will just get the `TypeError`

File: src/RequireCallStack.hs (hunk 1 above)
Comment (Gabriella439) [reply src/RequireCallStack.hs]: This is invoked in two situations:

- the user defines a (top-level or local) binding without a type annotation whose inferred type would have required `RequireCallStack`

  In other words, if the user does something like:
  
  ```haskell
  main = do
      let myError = errorRequireCallStack "oops!"
      
      provideCallStack (fromMaybe myError (Just (print 1)))
  ```
  
  … then that would tell the user to add `RequireCallStack` to the context and if the user follows that instruction:
  
  ```haskell
  main = do
      let myError :: RequireCallStack => a
          myError = errorRequireCallStack "oops!"
      
      provideCallStack (fromMaybe myError (Just (print 1)))
  ```
  
  … then the type error disappears.

- The user defines a (top-level or local) binding that has a type annotation but it is missing a `RequireCallStack`

  In other words, if the user were to do:
  
  ```haskell
  main = do
      let myError :: a
          myError = errorRequireCallStack "oops!"
      
      provideCallStack (fromMaybe myError (Just (print 1)))
  ```
  
  … then they would get the same error message telling them to add a `RequireCallStack` constraint.

File: src/RequireCallStack.hs [hunk 2]
```diff
@@ -48,9 +52,43 @@ import RequireCallStack.Internal
...
+data CallStack = CallStack
+
+-- | Raise an 'Control.Exception.ErrorCall' and incur a 'RequireCallStack'
+-- constraint while you do so. This
 --
 -- @since 0.1.0.0
 errorRequireCallStack :: RequireCallStack => String -> x
 errorRequireCallStack = error
+
+instance TypeError ('Text "Add RequireCallStack to your function context or use provideCallStack") => IP "provideCallStack" CallStack
```
Comment (hw202207) [inline src/RequireCallStack.hs]: Could you help to understand this line and how it connects to the fact that the error (`Add RequireCallStack to ...`) will be report if missing `RequireCallStack` constraint ?

File: src/RequireCallStack.hs [hunk 3]
```diff
@@ -48,9 +52,43 @@ import RequireCallStack.Internal
...
--- you do so. This
+-- | If you're running into this class, then you need to add
+-- 'RequireCallStack' to your function's signature, or discharge the
+-- constraint using 'provideCallStack'.
+--
+-- @since 0.1.0.0
+type RequireCallStackImpl = ?provideCallStack :: CallStack
+
+-- | The constructor for this type is intentionally not exported
+data CallStack = CallStack
```
Comment (hw202207) [inline src/RequireCallStack.hs]: nit: 
would you consider naming it slightly different such that won't be confused with the `GHC.Stack.CallStack`. 
`Confused` means when looking at code `?provideCallStack :: CallStack`, the first impression is the `CallStack` type is `GHC.Stack.CallStack` but later realized it is this `CallStack`.

File: src/RequireCallStack.hs (hunk 2 above)
Comment (Gabriella439) [reply src/RequireCallStack.hs]: I'm not sure I understand the question, but this error message will be reported whenever the user has a missing `RequireCallStack` constraint.  This code will remind them to add it or to use `provideCallStack` to eliminate the constraint if they don't want to keep propagating the callstack.

**Comment by parsonsmatt** [comment]
Upladed as 0.2.0.0, thank you!
